pkg/services/govmomi/power_test.go

```diff
@@ -0,0 +1,347 @@
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "vsphereVM1",
+					Namespace: "my-namespace",
+				},
+				Spec: infrav1.VSphereVMSpec{
+					VirtualMachineCloneSpec:  infrav1.VirtualMachineCloneSpec{},
+					PowerOffMode:             vmopv1.VirtualMachinePowerOpModeTrySoft,
+					GuestSoftPowerOffTimeout: &metav1.Duration{Duration: 1 * time.Minute},
+				},
+				Status: infrav1.VSphereVMStatus{
+					Conditions: []clusterv1.Condition{
+						{
+							Type:               infrav1.GuestSoftPowerOffCondition,
+							Status:             infrav1.GuestSoftPowerOffInProgressReason,
+							LastTransitionTime: metav1.NewTime(time.Now().Add(-2 * time.Minute)),
+						},
+					},
+				},
+			}
+
+			pending, err := vms.triggerSoftPowerOff(vmCtx)
+			g.Expect(pending).To(BeFalse())
+			g.Expect(err).ToNot(HaveOccurred())
+			return nil
+		})
+	})
+	t.Run("should return pending when VMware Tools is not running and powerOffMode set to soft", func(t *testing.T) {
+		g = NewWithT(t)
+		before()
+
+		simulator.Run(func(ctx goctx.Context, c *vim25.Client) error {
+			finder := find.NewFinder(c)
+			vm, err := finder.VirtualMachine(ctx, "DC0_H0_VM0")
+
+			vmCtx.Obj = vm
+			vmCtx.VSphereVM = &infrav1.VSphereVM{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "vsphereVM1",
+					Namespace: "my-namespace",
+				},
+				Spec: infrav1.VSphereVMSpec{
+					VirtualMachineCloneSpec:  infrav1.VirtualMachineCloneSpec{},
+					PowerOffMode:             vmopv1.VirtualMachinePowerOpModeSoft,
+					GuestSoftPowerOffTimeout: nil,
+				},
+			}
+
+			pending, err := vms.triggerSoftPowerOff(vmCtx)
+			g.Expect(pending).To(BeTrue())
+			g.Expect(err).ToNot(HaveOccurred())
+			return nil
+		})
+	})
+	t.Run("should return no pending when VMware Tools is not running and powerOffMode set to trySoft", func(t *testing.T) {
+		g = NewWithT(t)
+		before()
+
+		simulator.Run(func(ctx goctx.Context, c *vim25.Client) error {
+			finder := find.NewFinder(c)
+			vm, err := finder.VirtualMachine(ctx, "DC0_H0_VM0")
+
+			vmCtx.Obj = vm
+			vmCtx.VSphereVM = &infrav1.VSphereVM{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "vsphereVM1",
+					Namespace: "my-namespace",
+				},
+				Spec: infrav1.VSphereVMSpec{
+					VirtualMachineCloneSpec:  infrav1.VirtualMachineCloneSpec{},
+					PowerOffMode:             vmopv1.VirtualMachinePowerOpModeTrySoft,
+					GuestSoftPowerOffTimeout: &metav1.Duration{Duration: 1 * time.Minute},
+				},
+				Status: infrav1.VSphereVMStatus{
+					Conditions: []clusterv1.Condition{
+						{
+							Type:               infrav1.GuestSoftPowerOffCondition,
+							Status:             infrav1.GuestSoftPowerOffInProgressReason,
+							LastTransitionTime: metav1.NewTime(time.Now().Add(-2 * time.Minute)),
+						},
+					},
+				},
+			}
+
+			pending, err := vms.triggerSoftPowerOff(vmCtx)
+			g.Expect(pending).To(BeFalse())
+			g.Expect(err).ToNot(HaveOccurred())
+			return nil
+		})
+	})
+	// TODO: add more tests on VMware Tools reports running
```

Comment:
Kudos on the test coverage so far! Would be nice if we can add the remaining cases

Reply:
I could get the VMware tools reporting running simply.
I'm guessing some changes to the govmomi library would be needed.